util/fetch_shas.sh

```diff
@@ -2,17 +2,21 @@
 
 # Enumerates the list of expected downloadable files, loads the SHAs for each file, then
 # dumps the result to //rust:known_shas.bzl
-#
-# Should be run from workspace root.
 
 export LC_ALL=C
 
-TOOLS="$(cat ./util/fetch_shas_TOOLS.txt)"
-TARGETS="$(cat ./util/fetch_shas_TARGETS.txt)"
-VERSIONS="$(cat ./util/fetch_shas_VERSIONS.txt)"
-BETA_ISO_DATES="$(cat ./util/fetch_shas_BETA_ISO_DATES.txt)"
-NIGHTLY_ISO_DATES="$(cat ./util/fetch_shas_NIGHTLY_ISO_DATES.txt)"
-RUSTFMT_VERSIONS="$(cat ./util/fetch_shas_RUSTFMT_VERSIONS.txt)"
+# Detect workspace root
+if [[ -z "${BUILD_WORKSPACE_DIRECTORY}" ]]; then
+    SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+    BUILD_WORKSPACE_DIRECTORY="$( dirname "${SCRIPT_DIR}")"
+fi
+
+TOOLS="$(cat ${BUILD_WORKSPACE_DIRECTORY}/util/fetch_shas_TOOLS.txt)"
```

Comment:
Now that you're using a variable in them, can you quote all of the args to cat in this block?

Reply:
Updated